pkg/sources/sources.go

```diff
@@ -104,39 +104,81 @@ func NewConfig(opts ...func(*Config)) Config {
 	return *c
 }
 
+// Counter is a simple monotonically-increasing counter for tracking progress.
+// It includes a successCnt that records the number of successful scans.
+// It also includes a totalCnt that records the total number of scans.
+// This is necessary in order to display the number of concurrent scans
+// regardless of whether they are successful or not.
+// successCnt is the only field used when calculating the percentage complete.
+type Counter struct {
+	successCnt, totalCnt uint32
+	mu                   sync.Mutex
+}
+
+// IncTotal increments the total field of counter and returns the new value.
+func (c *Counter) IncTotal() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.totalCnt++
+	return int(c.totalCnt)
+}
```

Comment:
**Suggestion (optional):** I think the `atomic` package can be used here, particularly [AddUint32](https://pkg.go.dev/sync/atomic#AddUint32) which atomically adds to the counter and returns the new value.

I marked it at optional because I don't think it matters too much in this case to have a lock free solution.

Reply:
I started with that as well, but thought this slighlty more verbose way was a lot easier to read.

Reply:
As maybe a follow up question for you. in the `Update` method we need to get `successCnt` from the `Counter` would atomic also work in that case? I'm actually not super familiar with using atomic for anything other than incrementing a value